sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,652 @@
+    }
+
+    @Override
+    public void close() {
+      if (resultSet != null) {
+        try {
+          resultSet.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+      if (statement != null) {
+        try {
+          statement.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public Object getCurrent() {
+      return current;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write to a JDBC datasource.
+   */
+  public static class Write extends PTransform<PCollection<JdbcDataRecord>, PDone> {
+
+    public Write withDataSource(DataSource dataSource) {
+      return new Write(writer.withDataSource(dataSource));
+    }
+
+    public Write withUsername(String username) {
+      return new Write(writer.withUsername(username));
+    }
+
+    public Write withPassword(String password) {
+      return new Write(writer.withPassword(password));
+    }
+
+    private final JdbcWriter writer;
+
+    private Write(JdbcWriter writer) {
+      this.writer = writer;
+    }
+
+    @Override
+    public PDone apply(PCollection<JdbcDataRecord> input) {
+      input.apply(ParDo.of(writer));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<JdbcDataRecord> input) {
+      writer.validate();
+    }
+
+    private static class JdbcWriter extends DoFn<JdbcDataRecord, Void> {
+
+      private final DataSource dataSource;
+      private final String username;
+      private final String password;
+
+      private Connection connection;
+
+      public JdbcWriter(DataSource dataSource, String username, String password) {
+        this.dataSource = dataSource;
+        this.username = username;
+        this.password = password;
+      }
+
+      public JdbcWriter withDataSource(DataSource dataSource) {
+        return new JdbcWriter(dataSource, username, password);
+      }
+
+      public JdbcWriter withUsername(String username) {
+        return new JdbcWriter(dataSource, username, password);
+      }
+
+      public JdbcWriter withPassword(String password) {
+        return new JdbcWriter(dataSource, username, password);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+      }
+
+      @Setup
+      public void connectToDatabase() throws Exception {
+        if (username != null) {
+          connection = dataSource.getConnection(username, password);
+        } else {
+          connection = dataSource.getConnection();
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        JdbcDataRecord record = context.element();
```

Comment:
I don't understand what you mean. `JdbcDataRecord` is a container to wrap table name, column name, column type and column value. In the case of the sink, it's used to build the insert statement.